niworkflows/interfaces/images.py

```diff
@@ -188,6 +191,134 @@ def _run_interface(self, runtime):
         return runtime
 
 
+class _RobustAverageInputSpec(BaseInterfaceInputSpec):
+    in_file = File(
+        exists=True, mandatory=True, desc="A 4D file to average through the last axis"
+    )
+    t_mask = traits.List(traits.Bool, desc="List of selected timepoints to be averaged")
+    mc_method = traits.Enum(
+        None,
+        "AFNI",
+        "FSL",
+        usedefault=True,
+        desc="Which software to use to perform motion correction",
+    )
+    nonnegative = traits.Bool(
+        True, usedefault=True, desc="whether the output should be clipped below zero"
+    )
+
+
+class _RobustAverageOutputSpec(TraitedSpec):
+    out_file = File(exists=True, desc="the averaged image")
+    out_volumes = File(exists=True, desc="the volumes selected that have been averaged")
+    out_drift = traits.List(
+        traits.Float, desc="the ratio to the grand mean or global signal drift"
+    )
+    out_hmc = OutputMultiPath(File(exists=True), desc="head-motion correction matrices")
+
+
+class RobustAverage(SimpleInterface):
+    """Robustly estimate an average of the input."""
+
+    input_spec = _RobustAverageInputSpec
+    output_spec = _RobustAverageOutputSpec
+
+    def _run_interface(self, runtime):
+        img = nb.load(self.inputs.in_file)
+
+        # If reference is 3D, return it directly
+        if img.dataobj.ndim == 3:
+            self._results["out_file"] = self.inputs.in_file
+            self._results["out_volumes"] = self.inputs.in_file
+            self._results["out_drift"] = [1.0]
+            return runtime
+
+        fname = partial(fname_presuffix, self.inputs.in_file, newpath=runtime.cwd)
+
+        # Slicing may induce inconsistencies with shape-dependent values in extensions.
+        # For now, remove all. If this turns out to be a mistake, we can select extensions
+        # that don't break pipeline stages.
+        img.header.extensions.clear()
+        img = nb.squeeze_image(img)
+
+        # If reference was 4D, but single-volume - write out squeezed and return.
+        if img.dataobj.ndim == 3:
+            self._results["out_file"] = fname(suffix="_squeezed")
+            img.to_filename(self._results["out_file"])
+            self._results["out_volumes"] = self.inputs.in_file
+            self._results["out_drift"] = [1.0]
+            return runtime
+
+        img_len = img.shape[3]
+        t_mask = (
+            self.inputs.t_mask if isdefined(self.inputs.t_mask) else [True] * img_len
+        )
+
+        if len(t_mask) != img_len:
+            raise ValueError(
+                f"Image length ({img_len} timepoints) unmatched by mask ({len(t_mask)})"
+            )
+
+        n_volumes = np.sum(self.inputs.t_mask)
```

Comment:
what if this is undefined?
```suggestion
        n_volumes = np.sum(t_mask)
```

Reply:
Yup, this was the idea - thanks for catching it.